src/python/gudhi/wasserstein/quantization.py

```diff
@@ -0,0 +1,189 @@
+
+    .. note:: The exact algorithm presented in the reference paper
+              requires more technical considerations in order to prove theoretical convergence rates.
+              A theoretically motivated value (asymptotically optimal) for the ``batch_size`` is
+              ``int(log(len(pdiagset)))``.
+              However, numerical experiments suggest that this simplified version yields substantially similar results.
+    """
+
+    # We take the diagonal into account in the following.
+    # This could be turned into a parameter of the algorithm (so that it would encompass the standard
+    # "online-Lloyd algorithm"), but I thought this may be confusing.
+    withdiag = True
+
+    if not pdiagset:
+        raise ValueError("Input pdiagset is empty.")
+    # If no codebook has been provided, we propose an initialization.
+    if init is None:
+        init = _init_c(pdiagset, k)
+    # Number of centroids in the codebook (updated in the case c0 was provided by the user).
+    k = init.shape[0]
+
+    # Variable that will store the centroid evolution.
+    c_current = init.copy()
+
+    # number of diagrams
+    n = len(pdiagset)
+
+    # Building the batches (size batch_size, except possibly for the last one): a list of the form
+    # [[0,1,..., batch_size-1],[batch_size,..., 2batch_size-1],...,[remaining diags]]
+    # Note : we have to add (n % batch_size != 0) to increase the number of batches by 1 if needed.
+    nb_step = n // batch_size + (n % batch_size != 0)
+    batches = np.array_split(np.arange(0, n, dtype=int), nb_step)
+
+    # We now iterate through the diagrams
+    for t in range(nb_step):
+        # Get the empirical expected persistence diagram on the current batch
+        X_bar_t = _from_batch(pdiagset, batches[t])
+        # We compute the corresponding cells
+        cells = _get_cells(X_bar_t, c_current, withdiag=withdiag, internal_p=internal_p)
+
+        if order == 2.:
+            # We update the position of each centroid.
+            for j in range(k):
+                lc1 = len(cells[j])
+                # The batch may be empty (if c_current[j] does not catch any point)
+                # In which case we do not update it.
+                if lc1 > 0:
+                    # We compute the "gradient", i.e. the optimal descent direction for c_current[j]
+                    # When order == 2, it just consists of pushing c_current[j] towards the barycenter
+                    # of the cells[j] it defines.
+                    grad = np.mean(c_current[j] - cells[j], axis=0)
+                    # We then apply the gradient, with a 1/(1+t) factor to guarantee the convergence of this
+                    # stochastic-gradient-descent like approach (decreasing learning rate).
+                    c_current[j] = c_current[j] - grad / (t + 1)
+        else:
+            raise NotImplemented('Order = %s is not available yet. Only order=2. is valid' %order)
```

Comment:
I think you could error out earlier (or not provide this option at all and just say that it is W2).